### PR TITLE
High(er) performance `tryTakeN` for `Queue.bounded`

### LINKED ISSUE #2890: Implement optimized `tryTakeN` for async `Queue.bounded`
Following up from #2885, it should be possible to implement a very fast `tryTakeN` by porting `drain` from the jctools `MpmcArrayQueue`. This should result in a pretty dramatic optimization of Fs2's `Channel`, which is at the heart of a lot of programs now.

### PR DESCRIPTION
**Pull request by djspiewak**
Ports the `drain` function from the jctools `MpmcArrayQueue`. This basically just fast-paths the single-consumer bulk consumption scenario (which is to say, Fs2's `Channel`).

Builds on #2885 
Closes #2890

### CODE REVIEW

**Comment by djspiewak** [comment]
Depends on #3000

File: tests/shared/src/test/scala/cats/effect/std/QueueSpec.scala [hunk 1]
```diff
@@ -124,7 +124,7 @@ class BoundedQueueSpec extends BaseSpec with QueueTests[Queue] {
     }
 
     "offer/take at high contention" in real {
-      val size = 100000
+      val size = 10000
```
Comment (armanbilge) [inline tests/shared/src/test/scala/cats/effect/std/QueueSpec.scala]: If you want to keep it big for JVM.

```suggestion
      val size = if (cats.platform.Platform.isJVM) 100000 else 10000
```

Comment (djspiewak) [reply tests/shared/src/test/scala/cats/effect/std/QueueSpec.scala]: Wait `cats.platform.Platform` is public?

Comment (armanbilge) [reply tests/shared/src/test/scala/cats/effect/std/QueueSpec.scala]: It comes from `cats-kernel-laws` and is `private[cats]`. Good for us.

**Pull request merged**